src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs

```diff
@@ -2,381 +2,313 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Diagnostics;
 using System.Runtime;
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
-using Internal.Runtime;
-
 namespace Internal.Runtime.CompilerHelpers
 {
     /// <summary>
-    /// Math helpers for generated code. The helpers marked with [RuntimeExport] and the type
-    /// itself need to be public because they constitute a public contract with the .NET Native toolchain.
+    /// Math helpers for generated code. The helpers here are referenced by the runtime.
     /// </summary>
-    internal static partial class MathHelpers
+    [StackTraceHidden]
+    internal static class MathHelpers
     {
+        private const double Int32MaxValueOffset = (double)int.MaxValue + 1;
+        private const double UInt32MaxValueOffset = (double)uint.MaxValue + 1;
+
+        [RuntimeExport("Dbl2IntOvf")]
+        public static int Dbl2IntOvf(double value)
+        {
+            // Note that this expression also works properly for val = NaN case
+            if (value is > -Int32MaxValueOffset - 1 and < Int32MaxValueOffset)
+            {
+                return (int)value;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [RuntimeExport("Dbl2UIntOvf")]
+        public static uint Dbl2UIntOvf(double value)
+        {
+            // Note that this expression also works properly for val = NaN case
+            if (value is > -1.0 and < UInt32MaxValueOffset)
+            {
+                return (uint)value;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [RuntimeExport("Dbl2LngOvf")]
+        public static long Dbl2LngOvf(double value)
+        {
+            const double two63 = Int32MaxValueOffset * UInt32MaxValueOffset;
+
+            // Note that this expression also works properly for val = NaN case
+            // We need to compare with the very next double to two63. 0x402 is epsilon to get us there.
+            if (value is > -two63 - 0x402 and < two63)
```

Comment:
```suggestion
            if (value is > Math.BitDecrement(-two63) and < two63)
```
Would this work here?

Reply:
Not a compile-time constant

Reply:
> Would this work here?

Nope. BitDecrement is not a JIT intrinsic, and it is too complicated to make the JIT inline in and evaluate it into a constant. 

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGATCAFdgAGxgMAsgAoAlBy49FxAOyTsGABYA6AEIBLDABEYYWPhgA7DFPJaADDIDcC7gF8aroA===